pkg/virtual/framework/rootapiserver/server.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rootapiserver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/warning"
+	componentbaseversion "k8s.io/component-base/version"
+
+	virtualcontext "github.com/kcp-dev/kcp/pkg/virtual/framework/context"
+)
+
+var (
+	errorScheme = runtime.NewScheme()
+	errorCodecs = serializer.NewCodecFactory(errorScheme)
+)
+
+func init() {
+	errorScheme.AddUnversionedTypes(metav1.Unversioned,
+		&metav1.Status{},
+	)
+}
+
+// RootAPIServer is only responsible for serving the APIs for the virtual workspace
+// at a given root path or root path family
+// It does NOT expose oauth, related oauth endpoints, or any kube APIs.
+type RootAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+func NewServer(c completedConfig, delegationTarget genericapiserver.DelegationTarget) (*RootAPIServer, error) {
+	delegateAPIServer := delegationTarget
+	for _, vw := range c.Extra.VirtualWorkspaces {
+		var err error
+		delegateAPIServer, err = vw.Register(vw.Name, c.Generic, delegateAPIServer)
```

Comment:
Do I get this right? Root API server now can take list of virtual workspaces servers and becomes routing layer to route to delegated child API servers?

Reply:
It was like that already before. What changes is that the plumbing allows adding to that list after the options have been completed, i.e. in the config or server layer. Before authorization was wired in a way that this list could not be changed anymore.